lib/engine/game/base.rb

```diff
@@ -2807,37 +2825,55 @@ def game_end_check_values
         self.class::GAME_END_CHECK
       end
 
-      def custom_end_game_reached?
-        false
+      def game_end_timing(reason)
+        self.class::GAME_END_CHECK[reason]
       end
 
       def game_end_check
-        triggers = {
-          all_closed: all_closed?,
-          bankrupt: bankruptcy_limit_reached?,
-          bank: @bank.broken?,
-          stock_market: @stock_market.max_reached?,
-          final_train: @depot.empty?,
-          final_phase: @phase&.phases&.last == @phase&.current,
-          custom: custom_end_game_reached?,
-        }.select { |_, t| t }
-
-        %i[immediate current_round current_or full_or one_more_full_or_set].each do |after|
-          triggers.keys.each do |reason|
-            if game_end_check_values[reason] == after
-              @final_turn ||= @turn + 1 if after == :one_more_full_or_set
-              return [reason, after]
-            end
-          end
+        return @game_end_trigger if self.class::GAME_END_LOCK_FIRST_REASON && @game_end_trigger
+
+        triggers = game_end_check_values.each_with_object([]) do |(reason, _after), t|
+          t << [reason, game_end_timing(reason)] if send("game_end_check_#{reason}?")
         end
+        return if triggers.empty?
 
-        nil
+        trigger = triggers.min_by { |_, after| self.class::GAME_END_TIMING_PRIORITY[after] }
+        game_end_modify_final_turns!(*trigger)
```

Comment:
Nit: modify to me seems like it changes the content of the turns, where this is more of a set operation, no?

Reply:
The games that override it don't just set `@final_turn`, it's more open-ended depending on the game rules:
https://github.com/tobymao/18xx/blob/319aefbf76669605191010971823ed2c5d412491/lib/engine/game/g_1822_ca/game.rb#L535-L537

https://github.com/tobymao/18xx/blob/319aefbf76669605191010971823ed2c5d412491/lib/engine/game/g_1866/game.rb#L1012-L1018

https://github.com/tobymao/18xx/blob/319aefbf76669605191010971823ed2c5d412491/lib/engine/game/g_18_royal_gorge/game.rb#L1037-L1039

I don't really like the name I landed on though. Maybe `game_end_set_final_turn!` is still fine even if the overrides aren't expected to literally set `@final_turn`?

Reply:
Going with `game_end_set_final_turn!`

Reply:
I think its good enough. 👍